Fixes: #3550 Add Hello-world JAVA example using mill

himanshumahajan138 opened the pull request:
Fixes: #3550 

Created ```Build.sc``` file with proper format
Created ```helloworld.apk``` that runs on android

code runs with command ```mill app.create```

All code Running but the problem is with ```android.jar``` file which needs to be present locally or remotely 

@lihaoyi I request you to please once check the changes i made in the build.sc file its running well with the local code where android.jar is present but the problem is with the dependency for android.jar 

Hope You Understand.

screenshot of running app
![WhatsApp Image 2024-09-23 at 20 10 34_4891dfd5](https://github.com/user-attachments/assets/c0d34baa-0b3a-4bae-a680-ef801da79639)

0xnm commented:
There are few things you should address here:

* `__build` folder shouldn't be committed
* signing key shouldn't be committed as well: it should either come from the user or generated on the fly for debug builds
* `targetSdkVersion` should be updated to the 35 to keep up with the latest Android API
* you should either run this example in the system image which already have Android SDK installed and all the licenses accepted, or download it and install the necessary components from your build script (former option is better, I believe)
* `.bat` files you are trying to run won't be executed in Linux environment

himanshumahajan138 commented:
> * you should either run this example in the system image which already have Android SDK installed and all the licenses accepted, or download it and install the necessary components from your build script (former option is better, I believe)

@0xnm as you said i have to use the system image but i think its for my local testing but what will happen when i have to run this code on a system where sdk is not installed then do we need to create a local.properties file for explicating the path or we need to create our build in such a way that it installs the android sdk on its own if not present

please explain me a little bit in detail

himanshumahajan138 commented:
@lihaoyi @0xnm @lefou 

Fixes: #3550 

All the Checks Passed and All the files are committed as per requirements 

Directory Structure:
Got it! Here’s the directory structure exactly in the format you provided:

```
example/
│
javallib/
│
hello-world/
│   AndroidManifest.xml
│   build.sc
└───java/
    └───com/
        └───helloworld/
            └───app/
            │   MainActivity.java
```

This matches your original format. You can use this as is in your `README.md`.
Create & Compile App using ``` Mill _.compile ``` 

Kindly Review the Changes Made and Merge the Pull Request.

Screenshot :
![WhatsApp Image 2024-09-24 at 11 15 29_207f2586](https://github.com/user-attachments/assets/640b0c47-eb48-4409-bb4c-b0a042e92cec)

lihaoyi commented:
@himanshumahajan138 I think this looks good as a first pass, but there's a bunch of work to be done to bring it in line with the rest of Mill's codebase. Some immediate action items that come to mind:

1. The example needs to be running in CI using the standard `example/` folder format. 
    * This can go in a new `example/javalib/android/`, register it in `example/package.mill`
    * It needs a `/** Usage */` block with CLI instructions on how this is to be used, that will be executed as part of the example
2. The example should be self-contained and download all necessary files (as much as possible) as part of the build. 
    * You can use `requests.get` in a target to download the various `android.jar`, `d8`, `apksigner`, etc. executables as documented in https://mill-build.org/mill/Bundled_Libraries.html#_requests_scala
3. Let's move the bulk of the non-app-specific logic into an `trait AndroidModule` in `scalalib/src/mill/javalib/android`, so that only app-specific logic would live in the `build.sc` file
    * Scaladoc should be done using `/** */` to properly get picked up by the documentation system
4. The relationship between `AndroidModule` and `JavaModule` should be made clear. Does `AndroidModule` just provide the post-compilation step from classfiles->dex? Does it also provide custom source layouts and source->classfile compilation, or can it piggy-pack on the stuff that `JavaModule` already provides? I suspect a lot of the stuff calling `javac` ourselves is redundant
5. You should not need a `build_path` to put things in, the `T.dest` of each task should be sufficient, which you can wrap in `PathRef(T.dest)` and return from each task so downstream tasks can access the files using `upstreamTask().path / "blahblahblah"
6. The `build.sc` file should be `build.mill` as following the new file extension
7. Should use the new `os.call((...))` syntax instead of `os.proc().call()`, and `path / "platforms/android-35/android.jar"` instead of `path / "platforms" / "android-35" / "android.jar"`
8. The various `.bloop` stuff probably shouldn't be committed

himanshumahajan138 commented:
> 3\. Let's move the bulk of the non-app-specific logic into an `trait AndroidModule` in `scalalib/src/mill/javalib/android`, so that only app-specific logic would live in the `build.sc` file

@lihaoyi 

I am facing problem, please help
 i created AndroidModule.scala file in ```mill/scalalib/src/mill/javalib/android/``` and build.sc in ```mill/example/javalib/android/```

But now when i am importing ```mill.javalib.android.AndroidModule``` in the build.sc then its giving error that AndroidModule doesnt exist i tried my whole day solving this reading the docs and all but still unable to figureout why this is happening 

@lefou @0xnm @lihaoyi please help me if possible...

himanshumahajan138 commented:
> > i created `AndroidModule.scala` file in `mill/scalalib/src/mill/javalib/android/`
> 
> I can't see this file is this PR.

Actually I didn't pushed coz if i try to push it will fail in checks

lefou commented:
> @lefou @0xnm @lihaoyi please help me if possible...

I fear it's not possible to help when we can't see anything. If you see some error, you need to show it! You should push your changes, so CI can catch it, too. That makes it reproducible and easy to look at.

himanshumahajan138 commented:
Fixes : #3550 

@lihaoyi @lefou @0xnm 

I Changed Whole Code According to the Requirements and Code is running Fully in Local but i dont know what is happening when i pushed the code for live checking 

Key Changes :

- Create ```mill/example/javalib/android/1-hello-world/build.mill``` file which extends the ```AndroidModule.scala``` Trait and creates android app and then proper /** Usage */ for CLI 
- Created ```mill/scalalib/src/mill/javalib/android/AndroidModule.scala``` file which contains code for creating android application and then provided proper documentation in scaladoc format
- Updated ```mill/example/package.mill``` for android folder 

```Please help me out i have not slept from last 2 days constantly to create proper code for #3550 issue and now when everything was working fine in my local its again got some issue with the live testing```

Please Help me out !!! 
@lihaoyi @lefou @0xnm

lihaoyi commented:
@himanshumahajan138 if you have not slept for the last two days, I will tell you straight that you should stop working on this and go sleep. Maybe you are able to complete the ticket and maybe you can't, but regardless we're not interested in you working yourself to the bone.

himanshumahajan138 commented:
@lihaoyi Heartly Appreciated your words, please once review my pr and tell me what's wrong in this i again pushed some changes and this time i have followed each and every step as per requirements

himanshumahajan138 commented:
Fixes : #3550 

@lihaoyi @lefou @0xnm Please Review !!!

All test Passed and Code Worked as per the Requirements please Review and Merge #3596 to main branch

Finally after a Week of hustle 💯

lihaoyi commented on scalalib/src/mill/javalib/android/AndroidModule.scala:
```diff
@@ -0,0 +1,114 @@
+   * Directory where the Android SDK will be installed.
+   * Default is the 'android-sdk' directory within the Mill source path.
+   */
+  val sdkDir: os.Path = millSourcePath / "android-sdk"
```
You should not be installing things in random directories. You should define a target that downloads files to `T.dest`

himanshumahajan138 replied:
ya i will fix this will the help of this type of code 
`
def buildToolsPath: T[PathRef] = T {
    PathRef(sdkDir().path / "build-tools" / buildToolsVersion)
  }
  `

lihaoyi commented on example/javalib/android/1-hello-world/build.mill:
```diff
@@ -0,0 +1,137 @@
+    PathRef(genDir)
+  }
+
+  def compileJava = T {
+    val objDir = T.dest / "obj"
+    os.makeDir.all(objDir)
+    os.proc(
+      "javac",
+      "-classpath", androidJarPath,
+      "-d", objDir,
+      os.walk(generateResources().path).filter(_.ext == "java"),
+      os.walk(currentFolder / "java").filter(_.ext == "java")
+    ).call()
+    PathRef(objDir)
+  }
+
+  def createJar = T {
+    val jarPath = T.dest / "my_classes.jar"
+    os.proc(
+      d8,
+      os.walk(compileJava().path).filter(_.ext == "class"),  
+      "--output", jarPath,
+      "--no-desugaring"  
+    ).call()
+    PathRef(jarPath)
+  }
```
This should not be necessary to do ourselves, we should use the standard `JavaModule` mechanisms to do this

himanshumahajan138 replied:
we can extend JavaModule but we have to do these tasks like this to insure proper application integration with further steps

lihaoyi commented on example/javalib/android/1-hello-world/build.mill:
```diff
@@ -0,0 +1,137 @@
+      androidJarPath,
+      "--output", dexPath
+    ).call()
+    PathRef(dexPath)
+  }
+
+  def createApk = T {
+    val apkPath = T.dest / s"${appName}.unsigned.apk"
+    os.proc(
+      aapt,
+      "package",
+      "-f",
+      "-M", currentFolder / "AndroidManifest.xml",
+      "-I", androidJarPath,
+      "-F", apkPath,
+      createDex().path
+    ).call()
+    PathRef(apkPath)
+  }
+
+  def alignApk = T {
+    val alignedApkPath = T.dest / s"${appName}.aligned.apk"
+    os.proc(
+      zipalign,
+      "-f",
+      "-p",
+      "4",
+      createApk().path,
+      alignedApkPath
+    ).call()
+    PathRef(alignedApkPath)
+  }
+
+  def createKeystore = T {
+    val keystorePath = T.dest / "keystore.jks"
+    if (!os.exists(keystorePath)) {
+      os.proc(
+        "keytool",
+        "-genkeypair",
+        "-keystore", keystorePath,
```
None of this is unique to each application and should be in the shared `trait AndroidModule`

himanshumahajan138 replied:
you means all the application creation related code will go to to the `AndroidModule.scala`

lihaoyi replied:
`AndroidModule.scala` should have all the default configuration that applies to every application. `example/javalib/android/1-hello-world/` should only contain the configuration that is unique to this specific application, which for a hello-world project I expect to be very little

lihaoyi commented on scalalib/src/mill/javalib/android/AndroidModule.scala:
```diff
@@ -0,0 +1,114 @@
+      os.proc("unzip", zipFile, "-d", toolsDir).call()
+      val licensesDir: os.Path = sdkDir / "licenses"
+      os.makeDir.all(licensesDir)
+      os.write(licensesDir / "android-sdk-license", "24333f8a63b6825ea9c5514f83c2829b004d1fee")
```
What's this magic string?

himanshumahajan138 replied:
this is license key for installing the android sdk tools

himanshumahajan138 replied:
this is same for all the android sdk tools and is used to accept the licenses

lihaoyi replied:
This should be documented in a comment

himanshumahajan138 replied:
basically its developer's way to install android sdk otherwise there is no any other way

lihaoyi replied:
yes that's fine, it just needs to be documented in the source code in a comment so people know what this magic string is

lihaoyi commented on scalalib/src/mill/javalib/android/AndroidModule.scala:
```diff
@@ -0,0 +1,114 @@
+
+  /**
+   * Version of the Android build tools to be used.
+   */
+  val buildToolsVersion: String = "35.0.0"
+
+  /**
+   * Android platform version that will be targeted.
+   */
+  val platformVersion: String = "android-35"
+
+  /**
+   * Path to the Android build tools, based on the specified build tools version.
+   */
+  val buildToolsPath: os.Path = sdkDir / "build-tools" / buildToolsVersion
+
+  /**
+   * Path to the `android.jar` for the specified Android platform.
+   */
+  val androidJarPath: os.Path = sdkDir / "platforms" / platformVersion / "android.jar"
+
+  /**
+   * Path to the D8 Dex compiler, used for converting Java bytecode into Dalvik bytecode.
+   */
+  val d8: os.Path = buildToolsPath / "d8"
+
+  /**
+   * Path to the Android Asset Packaging Tool (AAPT), used for packaging Android apps.
+   */
+  val aapt: os.Path = buildToolsPath / "aapt"
+
+  /**
+   * Path to the Zipalign tool, used to optimize APKs for performance.
+   */
+  val zipalign: os.Path = buildToolsPath / "zipalign"
+
+  /**
+   * Path to the APK Signer tool, used for signing Android APK files.
+   */
+  val apksigner: os.Path = buildToolsPath / "apksigner"
```
All of these should be tasks, and those that return `os.Path` should return `PathRef`

himanshumahajan138 replied:
you means we have to use create like this 
```
def buildToolsPath: T[PathRef] = T {
    PathRef(sdkDir().path / "build-tools" / buildToolsVersion)
  }
```

lihaoyi replied:
yes

lihaoyi commented on example/javalib/android/1-hello-world/build.mill:
```diff
@@ -0,0 +1,137 @@
+  }
+
+
+  def compile = T {
+    if (!os.exists(currentFolder)) {
+      throw new java.io.IOException(s"Path $currentFolder does not exist!")
+    }
+    createApp() // Create the APK as part of the compilation process
+  }
```
`compile` should just be the standard `JavaModule#compile`, people who want to call `createApp` should just call `createApp`

himanshumahajan138 replied:
you mean we will add only one command ```./mill app.createApp``` instead of ```app.compile```

lihaoyi commented on scalalib/src/mill/javalib/android/AndroidModule.scala:
```diff
@@ -0,0 +1,114 @@
+  def installAndroidSdk(): Unit = {
+    // Step 1: Download the Android SDK command-line tools
+    val zipFile: os.Path = os.pwd / "commandlinetools-linux.zip"
+    if (!os.exists(zipFile)) {
```
When this is inside a target, these `os.exists` checks are not necessary

himanshumahajan138 replied:
okie i will remove this code

lihaoyi commented on example/javalib/android/1-hello-world/build.mill:
```diff
@@ -0,0 +1,137 @@
+    os.proc(
+      aapt,
+      "package",
+      "-f",
+      "-M", currentFolder / "AndroidManifest.xml",
+      "-I", androidJarPath,
+      "-F", apkPath,
```
Are there long flags we can use here?

himanshumahajan138 replied:
yes there are available long flags also do we need to use ?

lihaoyi replied:
yes

lihaoyi commented:
I've updated the original ticket with some more details. I think you've made good progress, but still a ways to go

* [ ] A folder in `example/javalib/android/1-hello-world` that contains a build that contains Java code for building a trivial hello-world android app APK, that can then be loaded into a physical phone or android emulator. 
    * This isn't ready yet, the build is not yet of acceptable quality to be merged as an official example
* [x] Just the simplest plain-old-Java android app is fine: no need fancier features or C++ code or whatever. 
* [x] Manual testing of the final APK is fine, but the process to build it should be exercised in the example. 
    * I assume you've tested it, and the example passes
* The app should have two modules containing Java code, one depending on another, to demonstrate how splitting of a app codebase into modules works
    * [ ] Not done yet. Currently it's only one module
* Android-specific config should be extracted into a `trait AndroidJavaModule` in `mill.javalib`, and only app-specific config should be in the `example/javalib/android/1-hello-world` example
    * [ ] This is not yet of acceptable quality to be merged
* Both the library code and the example code should be rigorously documented w.r.t. how it is used and how it fits into the larger Android ecosystem
    * [ ] This is not there yet. Doc-comments are a start, but that does not achieve `rigorously documented w.r.t. how it is used and how it fits into the larger Android ecosystem`

In general, the code now works, but for it to be an official Mill-android example we need to aim for higher than "works". Please go through all the other examples in `example/`, especially `example/javalib/` `example/scalalib/` and `example/kotlinlib/`, along with their associated `scalalib/` and `kotlinlib/` traits to see what a Mill example and module trait should look like. The example code doesn't need to be perfect, but it needs to fit into the prevailing Mill style and coding conventions, which this PR currently still does not

himanshumahajan138 commented:
> The app should have two modules containing Java code, one depending on another, to demonstrate how splitting of a app codebase into modules works
> * [ ]  Not done yet. Currently it's only one module

You Means we have to create two functions inside `build.mill`  
or 
Create two Separate Traits one for Downloading Android Setup and other for creating Android app and then extend them inside `build.mill`

lihaoyi commented:
> You Means we have to create two functions inside build.mill
or
Create two Separate Traits one for Downloading Android Setup and other for creating Android app and then extend them inside build.mill

Two traits sounds reasonable. Assuming we provide good defaults, each trait we extend inside build.mill will be somewhere around 1-3 lines of code, not a huge burden to someone configuring the build. We could also set it up such that for someone who only has one android app, they can extend both traits in the same module

lihaoyi commented:
Notably, two modules doesn't mean two android apps. It means two java modules that end up being included in the one same android app

himanshumahajan138 commented:
> > You Means we have to create two functions inside build.mill
> > or
> > Create two Separate Traits one for Downloading Android Setup and other for creating Android app and then extend them inside build.mill
> 
> Two traits sounds reasonable. Assuming we provide good defaults, each trait we extend inside build.mill will be somewhere around 1-3 lines of code, not a huge burden to someone configuring the build. We could also set it up such that for someone who only has one android app, they can extend both traits in the same module

OK so I will add One named as `AndroidModule.scala`  and other `AppModule.scala`

lihaoyi commented:
Let's name them `AndroidSdkModule.scala` and `AndroidAppModule.scala`, a bit more verbose but makes their roles clearer

himanshumahajan138 commented:
Fixes: #3550 
@lihaoyi 
Sir can you please review my `AndroidAppModule.scala` 
I have formatted it fully and it passed local linting check 
But i dont know why it failed maybe some thing left on my part which i am unable to figureout,
So i request you to please review the code ✅

himanshumahajan138 commented:
Fixes #3550 
@lihaoyi 

Final All test Passed and All Requirements also completed 
Please Review the code and then i will start documentation.....

@lihaoyi @0xnm @lefou

lihaoyi commented:
@himanshumahajan138 I'm going to make the call and say that this collaboration isn't going to work out. I don't think you have the expertise necessary to push this over the finish line, there's a limit in how much I can hold your hand, and I don't see this PR converging to a merge-able state with a reasonable amount of time and effort.

Thank you for the time you put in, I know you've tried hard and done your best, but I'm going to call it quits on this one and close this PR to open up the space for others to attack the problem

himanshumahajan138 commented:
@lihaoyi 
I think this is not fare ! 
All the things are almost covered
Keypoints:
  1. Created build.mill which inherits AndroidAppModule and AndroidSdkModule 
  2. AndroidSdkModule trait provides the implementation for Sdk download 
  3. AndroidAppModule trait provides the implementation for App Creation 
  4. App configuration related stuff will be updated in build.mill file making it unique from default app creation
  5. Proper scaladoc and usage followed to implement the code 
  6. App testing done in local, results passed 
  7. Code fully reusable for unique application
  8. Passed all automated testcases 
  
Just asking for review because i want to create documentation based on finetuned code but when i was near finish line you just kicked me out 
  i heartly request you to please tell me what is left and i will provide you with the professional code and documentation 
  i have worked a lot for this please understand....
  please give me a chance !
